### PR TITLE
Support for masking (variable length input)

### PR DESCRIPTION
**Pull request by vgire**
This is a PR for https://github.com/Element-Research/rnn/issues/45
(Sorry for the name change, @sensout is an organization).

I've tested it on both CPU and GPU (CUDA only).
See my remarks about mandatory forward padding.
You might find a more generic and elegant way to cope with the different `recurrentModule` signatures.

About performance, I've seen a 10% speed decrease with zero mask enabled.
Checking all inputs to find even a few paddings is costly (there might be better options than computing the norm). It would be nice if performance penalty was proportional to the number of paddings only (but I don't see how).

Note: I'm currently using it with a LookupTable where the weight is forced to zero for padding and I guess I could also make a PR for a LookupTableZeroMask (or whatever name you prefer) inheriting `nn.Decorator` to support any LookupTable implementation.

Cheers

### CODE REVIEW

**Comment by nicholas-leonard** [comment]
@vgire So that was the fastest, best looking PR I have seen in a while. I modified it to make it work with table inputs and outputs. To do so, I had to add the `nInputDim` argument to the constructor. If you see anything I missed, let me know. Thanks for the quick and awesome PR. Keep 'em coming!

**Comment by vgire** [comment]
Thanks ! Easy when underlying implementation was made for it ;)
I had not realized that inputs and outputs could be tables.
Also great that you spot the `gradOutputs` zeroing bug. 
I PRed some minor fixes.

**Comment by vzhong** [comment]
Hey @vgire , would you mind doing a PR for the lookup table as well?

**Comment by vgire** [comment]
Hey @vzhong,
For now, I am just zeroing the `nn.LookupTable` weight row for padding with `lookupTable.weight[padIndex]:zero()` before forwarding batches.
This is a bit of a hack but it is fast and works well if the LookupTable is the first layer of your network (i.e. you don't care about gradInputs).

I have to find a cleaner (but fast) approach to do a PR. 
I am thinking of implementing a `nn.Decorator` which would work in a more general case (`nn.LookupTable`, `nn.Linear` with OneHot) but not sure yet how to implement this (especially computing the gradInputs) without too much performance penalty.
@nicholas-leonard what do you think ?

**Comment by nicholas-leonard** [comment]
@vgire I would just extend the LookupTable module by adding a `zeroEmbedding(index)` method that basically overrides the `updateOutput` by making the new such method do your thing (`lookupTable.weight[padIndex]:zero()`) before calling the old method. You could add this to dpnn in a LookupTable.lua file. Later, each module could be extended in its own way, by allow a call to the same method `zeroEmbedding()`.

**Comment by vzhong** [comment]
So I ended up hacking together a LookupTable that supports masking:

```lua
require 'nngraph'
local MaskedLookupTable = torch.class('MaskedLookupTable')

function MaskedLookupTable:__init(n_vocab, n_emb)
  local mask = nn.Identity()()
  local indices = nn.Identity()()
  local input = {indices, mask}
  local lookup = nn.LookupTable(n_vocab, n_emb)
  local embed = lookup(indices)
 
  -- mask needs to be repeated n_emb number of times
  -- the 3rd dimension is the embedding dim
  local repeated_m_mask = nn.Replicate(n_emb, 3)(mask)
  local masked_embed = nn.CMulTable(){embed, repeated_m_mask}
  local output = {masked_embed}
 
  self.lookup = lookup
  self.module = nn.gModule(input, output)
end

return MaskedLookupTable
```

This has the downside of requiring a `mask` parameter as input, which it uses to zero out the embeddings corresponding to indices with mask zero.

**Comment by vgire** [comment]
For reference and @vzhong in case you are still looking for it, i did a PR for the LookupTable https://github.com/Element-Research/rnn/pull/54

**Comment by hughperkins** [comment]
Hi, just to check, does this zero the hidden state at the start of each sequence within the mini-batch?  Or it is like, you can feed in different-sized sequences in a mini-batch, and it will pad each sequence with zeros, so they become all the same length?

**Comment by vgire** [comment]
`MaskZero` or `LookupTableMaskZero` don't do the padding.
`MaskZero` ignore the inputs with zero norm and `LookupTableMaskZero` project a 0 index element to a zero norm vector which will be ignored by `MaskZero`.
You can indeed work with different-sized sequences in a mini-batch (that's the whole point) but you must pad each sequence with LEADING zeros yourself.
I hope this answers your question.

**Comment by hughperkins** [comment]
Ah, interesting.  Sounds good :-)

**Comment by nicholas-leonard** [comment]
@hughperkins I think you could use this functionality for your own purpose of forgetting. To do so, you need to pad the sequences with zeros such that the end of the sequences are aligned. Then you forget and continue on with the next batch of variable length sequences.

**Comment by hughperkins** [comment]
Yes, seems plausible.  Will check this point :-)
